clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3209,65 +3209,148 @@ static void handleWorkGroupSizeHint(Sema &S, Decl *D, const ParsedAttr &AL) {
+                                            const AttributeCommonInfo &CI,
+                                            Expr *E) {
   if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
     llvm::APSInt ArgVal;
-    ExprResult ICE = S.VerifyIntegerConstantExpression(E, &ArgVal);
-
-    if (ICE.isInvalid())
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
       return;
+    E = Res.get();
 
-    E = ICE.get();
-    int64_t NumSimdWorkItems = ArgVal.getSExtValue();
-
-    if (NumSimdWorkItems == 0) {
-      S.Diag(E->getExprLoc(), diag::err_attribute_argument_is_zero)
-          << AL << E->getSourceRange();
+    // This attribute requires a strictly positive value.
+    if (ArgVal <= 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
       return;
     }
 
-    if (const auto *A = D->getAttr<ReqdWorkGroupSizeAttr>()) {
-      ASTContext &Ctx = S.getASTContext();
-      Optional<llvm::APSInt> XDimVal = A->getXDimVal(Ctx);
-      Optional<llvm::APSInt> YDimVal = A->getYDimVal(Ctx);
-      Optional<llvm::APSInt> ZDimVal = A->getZDimVal(Ctx);
-
-      if (!(XDimVal->getZExtValue() % NumSimdWorkItems == 0 ||
-            YDimVal->getZExtValue() % NumSimdWorkItems == 0 ||
-            ZDimVal->getZExtValue() % NumSimdWorkItems == 0)) {
-        S.Diag(AL.getLoc(), diag::err_sycl_num_kernel_wrong_reqd_wg_size)
-            << AL << A;
-        S.Diag(A->getLocation(), diag::note_conflicting_attribute);
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelNumSimdWorkItemsAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue());
+      if (DeclExpr && ArgVal != DeclExpr->getResultAsAPSInt()) {
+        Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+        Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        return;
+      }
+    }
+
+    // If the declaration has an [[intel::reqd_work_group_size]] attribute,
+    // check to see if can be evenly divided by the num_simd_work_items attr.
+    if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+      Optional<llvm::APSInt> XDimVal = DeclAttr->getXDimVal(Context);
+      Optional<llvm::APSInt> YDimVal = DeclAttr->getYDimVal(Context);
+      Optional<llvm::APSInt> ZDimVal = DeclAttr->getZDimVal(Context);
+
+      if (!(*XDimVal % ArgVal == 0 || *YDimVal % ArgVal == 0 ||
+            *ZDimVal % ArgVal == 0)) {
```

Comment:
I understand this is current behavior but based on email discussion, required behavior is yet to be verified right? Will this be fixed done in a follow-up patch if required?

Reply:
Correct, whatever @smanna12 finds out about this should be handled in a follow-up, I think.